Qbeast-io/qbeast-spark
# Progressive append

Fixes #132. 

### Compress small trees when appending to an existing revision
Appending a small volume of data to an existing qbeast table is likely to create many small files, as shown in #132. To tackle the issue we can recursively group and send sibling `payloads` to their parent cubes as long as the resulting size is under a certain threshold, resulting in fewer and larger blocks to write.

The process is called COMPRESSION and is done after the `cubeColumnName` is created and before repartitioning data by cubes.

The process is only triggered when appending data to an existing revision, for it is of our interest for the initial tree to have a natural depth. On top of that, the appended data has to be small (<= `MAX_SIZE_FOR_APPEND_COMPRESSION`) otherwise blocks are already large. Compressing a large tree is expensive.

`MAX_SIZE_FOR_APPEND_COMPRESSION` is configurable through `spark.qbeast.index.maxAppendCompressionSize`, default value is set to 0.

### Influence on cube maxWeights
Cube `maxWeight`s from the parallel implementation **increase monotonically** from top to bottom on any branch. We need to make sure the proposed compression won't alter this behavior - when moving cube elements to their ancestors, both parent and child `maxWeight`s are left intact, in fact, the append won't even affect the `maxWeight` of the child cube since no elements are written, and the ancestor cube will have `maxWeight` reduced by the update.

### Influence on online sampling and querying
The compression doesn't modify the `maxWeight`s assigned during estimation. Branch `maxWeight`s are left monotonically increasing so it won't affect the correctness of our sampling. The efficiency, however, may drop for blocks now have more elements than that reflected by their `maxWeight`; the excess elements are later filtered in memory.

- [x] Add/update documentation
- [x] New feature / bug fix has been committed following the [Contribution guide](https://github.com/Qbeast-io/qbeast-spark/blob/main/CONTRIBUTING.md).
- [x] Add comments to the code (make it easier for the community!).
- [x] Your branch is updated to the main branch (dependent changes have been merged).

Tests:
- Make sure the compressed tree has fewer cubes than the original
- Make sure adding a compressed tree doesn't change the fact that maxWeights increase monotonically in any branch
- Make sure that inline sampling is not corrupted by adding a compressed tree

Test Configuration:
* Spark Version: `3.2.2`
* Hadoop Version: `3.3.1`
* Cluster or local? `Local`

TODO:
- [x] Compute cube sizes from delta and existing cube weights in order to avoid collect
- [x] Perform compression through roll-up and use the resulting map during indexing
- [x] Consider cube state during compression
